### PR TITLE
⚠️ Remove v1alpha3 and v1alpha4

### LINKED ISSUE #1326: remove v1alpha3/4 ?
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

we stated v1alpha3/4 is for CAPI v1alpahx version and now CAPI is already beta
should we remove the early support in main branch so that we can avoid maintain unnecessary complex code e.g conversions?

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by mdbooth**
**Which issue(s) this PR fixes**:
Fixes #1326

**Special notes for your reviewer**:
This PR relies on behaviour in clusterctl which updates storage versions on upgrade.

/hold

### CODE REVIEW

File: .golangci.yml
[the remark this answers is not in the document]
Comment (mdbooth) [reply .golangci.yml:1]: Added as a second commit.

**Review by mnaser — APPROVED** [review]

**Review by seanschneeweiss — COMMENTED** [review]
Keeping compatibility for that many revisions of the API of course is a pain and should be minimized. So thanks beforehand for working on this.

Two thoughts on this:
- Looking at CAPI, where they did almost the same for their Alpha versions. CAPI deprecated the API before removing it. https://github.com/kubernetes-sigs/cluster-api/pull/8071. Should we deprecate first and then remove - regardless the fact, that this is an alpha API version? From [src](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api):
  > Alpha API versions may be removed in any release without prior deprecation notice 
- Should we adapt https://github.com/kubernetes-sigs/cluster-api/issues/8038?
Especially this part
> - We will set `served` to `false` for the apiVersions in all our CRDs.
> - Then users can’t read or write with the old apiVersions anymore.
> - The API server will still be able to read and convert the old apiVersions.
> - If necessary, users can easily enable the apiVersions again by reverting `served` back to `true` on the CRD.

**Review by seanschneeweiss — COMMENTED** [review]
This might need modification too:
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/17d3f369e93a443c2711693eb14cd2a3d50483e0/config/crd/kustomization.yaml#L1-L3

**Comment by mdbooth** [comment]
> Keeping compatibility for that many revisions of the API of course is a pain and should be minimized. So thanks beforehand for working on this.
> 
> Two thoughts on this:
> 
> * Looking at CAPI, where they did almost the same for their Alpha versions. CAPI deprecated the API before removing it. [🌱 Deprecate v1alpha3 & v1alpha4 cluster-api#8071](https://github.com/kubernetes-sigs/cluster-api/pull/8071). Should we deprecate first and then remove - regardless the fact, that this is an alpha API version? From [src](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api):
>   > Alpha API versions may be removed in any release without prior deprecation notice
> * Should we adapt [Removing v1alpha3 & v1alpha4 apiVersions cluster-api#8038](https://github.com/kubernetes-sigs/cluster-api/issues/8038)?
>   Especially this part
> 
> > * We will set `served` to `false` for the apiVersions in all our CRDs.
> > * Then users can’t read or write with the old apiVersions anymore.
> > * The API server will still be able to read and convert the old apiVersions.
> > * If necessary, users can easily enable the apiVersions again by reverting `served` back to `true` on the CRD.

I think this is the important discussion, and I'm ambivalent.

I am inclined to say lets get these 2 specifically gone and do better in the future. The reasons are:
* We haven't effectively tested these versions in a long time. I have little confidence they still work.
* Anybody who upgraded to v0.7 is no longer running them anyway.
* Anybody who skips directly to v0.8 can recover by downgrading to v0.7, waiting for the upgrade to complete, then upgrading to v0.8 again. We need to ensure this is in the 0.8 release notes, though.
* We're making the continued modification of the current API more burdensome by keeping them.

However, I appreciate this is a bullish view and I could easily be persuaded otherwise.

In the spirit of 'do better in the future', though, I wonder if we should immediately deprecate v1alpha5 and v1alpha6 with a target of setting 'unserved' in v0.9 and removal in 'v0.10'?

**Comment by jichenjc** [comment]
>>>In the spirit of 'do better in the future', though, I wonder if we should immediately deprecate v1alpha5 and v1alpha6 with a target of setting 'unserved' in v0.9 and removal in 'v0.10'?

is this CAPI or other provider should consider already? we may not be identical to them but reference  should be helpful

anyway, I think your statement is reasonable and we want folks to keep upgrade to save effort with lower API version maintainance

**Comment by mdbooth** [comment]
I've added this to the agenda of this week's office hours. Lets make a decision on this by the end of this week at the latest.

**Review by seanschneeweiss — APPROVED** [review]

**Comment by jichenjc** [comment]
/lgtm

thanks for the update

**Pull request merged**